docs/investigations/secure-resolvers/secure-auth-resolvers.md

```diff
@@ -0,0 +1,184 @@
+We can restrict the access by current `hasScopes` directive.
+
+### Restrict access to resolvers for specific types of API consumers (by ID)
+
+We should be able to restrict access to any resolver that takes API Consumer ID as one of its parameters by adding directive.
+
+We want to cover following cases:
+
+1. Runtime musn't read other Runtimes.
+2. Application musn't read other Applications.
+3. Integration System musn't read other Integration Systems.
+4. Runtime musn't read Applications which are not in the same scenario (`applicationsForRuntime`).
+
+### Restrict access to Application/Runtime for IntegrationSystem is which not managed
+When Integration System request specific Application/Runtime, it should be able to read only object managed by itself.
+
+## Solution
+To achieve those restrictions the following solution is proposed:
+
+### Add information about API consumer in Tenant Mapping service
+Tenant Mapping Service adds 2 items to the JWT token
+1. Object Type, service is able to determine who calls the API, whether it is Application, Runtime or IntegrationSystem
+2. Object ID of caller
+
+The enriched JWT is sent to the Compass API.
+
+### Directive `limitAccessFor`
+This directive will limit access to resolvers for specific types of API consumers (by ID).
+
+Proposed directive:
+```graphql
+enum consumerType {
+    RUNTIME
+    APPLICATION
+    INTEGRATION_SYSTEM
+    USER
+}
+
+directive @limitAccessFor(consumerType: consumerType!, idField: String!) on FIELD_DEFINITION
+```
+
+The proposed directive does following things:
+1. get object type and ID from JWT token
+2. check if object is the same type as `consumerType` argument
+3. if yes, then the directive compare the ID field. If the ID mismatches, the following error is returned `Access Denied`.
+4. if no, the request is allowed
+
+Currently we cannot use this directive on query param: [issue](https://github.com/99designs/gqlgen/issues/760).
+
+### Directive `limitIntegrationSystemAccessTo`
+This directive will limit access to application/runtime resolvers managed by integration system
+
+Proposed directive:
+```graphql
+enum managedResourceType {
+    RUNTIME
+    APPLICATION
+}
+
+directive @limitIntegrationSystemAccessTo(managedResourceType: managedResourceType!, idField: String!) on FIELD_DEFINITION
+```
+
+The proposed directive does following things:
+1. get object type and ID from JWT token
+2. check if object is the `integration system`
+3. if yes, then the directive check in database if given application/runtime is managed by integration system which called the query.
+If integration system doesn't manage the application/runtime, the following error is returned `Access Denied`
+
+Currently we cannot use this directive on query param ( https://github.com/99designs/gqlgen/issues/760).
+
+### Database schema change
+Every resource managed by integration system have to have column with reference to integration system table record.
```

Comment:
Why this is necessary? I would say that we need to distinguish the ID of the Integration System via the managed object but I would not say that we need to have the exact column. It can be done for example via the technical table that joins the Application with the Integration System.

Reply:
I would write sth like this:
```
Every resource managed by integration system have to be reachable from integration system point of view.
```

Reply:
`point of view` => `perspective`